### PR TITLE
Add deploy while moving to Saboteur

### PR DESCRIPTION
**Pull request by PunkPun**
Enabling cloak should not be interfering with the current actors activity

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs [hunk 1]
```diff
@@ -46,6 +46,9 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
 		[Desc("Can " + nameof(DeployedCondition) + " be canceled by followup deploy order?")]
 		public readonly bool CanCancelCondition = false;
 
+		[Desc("Can the actor deploy while moving?")]
+		public readonly bool CanDeployWhileMoving = false;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]:
```suggestion
		public readonly bool CancelsCurrentOrders = false;
```
To be more generic as this doesn't specifically apply to movement.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs [hunk 2]
```diff
@@ -151,7 +154,15 @@ Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
 		public void ResolveOrder(Actor self, Order order)
 		{
 			if (order.OrderString == "DeployWithCharge")
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]:
```suggestion
			if (order.OrderString != "DeployWithCharge")
			    return;
```
For improved readability.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs [hunk 3]
```diff
@@ -151,7 +154,15 @@ Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
...
-				self.QueueActivity(order.Queued, new DeployForGrantedConditionWithCharge(self, this));
+			{
+				if (Info.CanDeployWhileMoving && !order.Queued)
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]: Invert `if` for improved readability please.

**Review by penev92 — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]: your suggestion is plain wrong though, we never cancel orders

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]: and this does specifically apply to movement

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs (hunk 3 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]: I find this less readable

Screenshot 2024-12-12 at 12 56 25

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeployWithCharge.cs]: renamed the trait to ToggleChargedConditionOnOrder

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Allow deploying on specified charge to grant a condition for a specified duration.")]
-	public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo, IRulesetLoaded
+	public class ToggleChargedConditionOnOrderInfo : PausableConditionalTraitInfo, IRulesetLoaded
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: I don't think we want the `OnOrder` here. All these traits act in response to an order thats generated by the trait itself. The distinction for `ToggleConditionOnOrder` is that does not issue the order, but responds to an order created by other unrelated code.

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 5]
```diff
@@ -46,6 +46,9 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
 		[Desc("Can " + nameof(DeployedCondition) + " be canceled by followup deploy order?")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: "deploy" terminology should be removed.

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 6]
```diff
@@ -46,6 +46,9 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
 		public readonly bool CanCancelCondition = false;
 
+		[Desc("Should we interrupt the current activity and wait for it to finish?")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]:
```suggestion
		[Desc("Should we interrupt the current activity if the toggle order is not queued?")]
```

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 7]
```diff
@@ -150,8 +153,13 @@ Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
...
 		{
-			if (order.OrderString == "DeployWithCharge")
-				self.QueueActivity(order.Queued, new DeployForGrantedConditionWithCharge(self, this));
+			if (order.OrderString != "DeployWithCharge")
+				return;
+
+			if (order.Queued || Info.CancelsCurrentOrders)
+				self.QueueActivity(order.Queued, new ToggleChargedCondition(self, this));
+			else if (CanDeploy())
+				TriggerDeploy(self);
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]:
```suggestion
				ToggleDeploy(self);
```
(or ToggleState or something since @pchote wanted to remove the "deploy" terminology)

**Review by penev92 — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: my version descried the behaviour correctly

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: charging is a concept in this traits name, I think its fine

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs (hunk 6 above)
Comment (penev92) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: I don't understand "interrupt and wait to finish". The point is to _not_ wait for it to finish.
Also that depends on whether the order is queued 🤷‍♂️

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 8]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
 
-		[Desc("Play a randomly selected sound from this list when deploying.")]
-		public readonly string[] DeploySounds = null;
+		[Desc("Play a randomly selected sound from this list when turning on.")]
+		public readonly string[] ActivateSounds = null;
 
-		[Desc("Play a randomly selected sound from this list when undeploying.")]
-		public readonly string[] UndeploySounds = null;
+		[Desc("Play a randomly selected sound from this list when turning off.")]
+		public readonly string[] DeactivateDounds = null;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: Typo:
```suggestion
		public readonly string[] DeactivateSounds = null;
```

Also can we go with `ActivationSounds`/`DeactivationSounds` please?

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 9]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
 
 		[VoiceReference]
 		public readonly string Voice = "Action";
 
-		public readonly Color ChargingColor = Color.Magenta;
-		public readonly Color DeployedColor = Color.DarkMagenta;
+		[Desc("Color to display when deactivated.")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: *Color of the charge bar?

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 10]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
+		public readonly Color DeactivatedColor = Color.Magenta;
+
+		[Desc("Color to display when activated.")]
+		public readonly Color ActivatedColor = Color.DarkMagenta;
+
+		[Desc("Display the bar when no charge is available or the trait is disabled.")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]:
```suggestion
		[Desc("Should the charge bar be displayed when not charged or the trait is disabled?")]
```

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs [hunk 11]
```diff
@@ -80,37 +89,37 @@ public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
...
 
-	public class GrantConditionOnDeployWithCharge : PausableConditionalTrait<GrantConditionOnDeployWithChargeInfo>,
+	public class ToggleChargedConditionOnOrder : PausableConditionalTrait<ToggleChargedConditionOnOrderInfo>,
 		IIssueOrder, IResolveOrder, ITick, ISelectionBar, IOrderVoice, ISync, IIssueDeployOrder
 	{
 		[Sync]
 		int chargeTick = 0;
 
-		bool deployed = false;
+		bool activated = false;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]:
```suggestion
		bool isActive = false;
```

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs (hunk 4 above)
Comment (penev92) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: I was looking at the updated PR's yaml portion and seeing "Toggle...OnOrder" with "ActiveCondition" below I was fully expecting to also see an order name (but ofc there isn't one). So that suffix can indeed be misleading and should probably go away.

**_HOWEVER_**, while glancing through the trait changes and specifically at `VoiceForOrder`, I got thinking that this will NOT work for more than one such trait per actor because they use the same order string. What if we solve both issues with a single fix - add an `Order` field to the TraitInfo and use that for issuing/resolving, _(For safety's sake maybe go with `$"ActivateCondition-{info.Order}"`?)_

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: You interrupt but the activity might still take quite a few frames to finish. We wait for it, instead of skipping like the default behaviour in this PR is

File: OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: having multiple enabled breaks most order traits, the key is that you can have multiple by disabling one and exanbling another

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/ToggleChargedConditionOnOrder.cs]: renamed to `GrantChargedConditionOnToggle`

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs [hunk 12]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
 
 		[Desc("How long (in ticks) should the condition stay active?")]
 		public readonly int ConditionDuration = 1;
 
-		[Desc("Can " + nameof(DeployedCondition) + " be canceled by followup deploy order?")]
+		[Desc("Can " + nameof(ActivatedCondition) + " be canceled turned off manually?")]
 		public readonly bool CanCancelCondition = false;
 
+		[Desc("Should we interrupt the current activity and wait for it to finish?")]
+		public readonly bool CancelsCurrentOrders = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs]: What is the use case of this flag? We have `GrantConditionOnDeploy` for things that "deploy" (i.e. cancel current orders). It would be better to keep the "OnDeploy" terminology for that use case and keep this trait separate.

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs [hunk 13]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
+
 		[CursorReference]
-		[Desc("Cursor to display when able to (un)deploy the actor.")]
-		public readonly string DeployCursor = "deploy";
+		[Desc("Cursor to display when able to trigger a state change.")]
+		public readonly string OrderCursor = "deploy";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs]: Why "Order"? We don't preface other types of cursors used by other traits.

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs (hunk 12 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs]: grant condition on deploy doesn't do charge, the usecase is the general flexibility conditions allow to modders. The condition may trigger something that needs the unit to be stationary

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs [hunk 14]
```diff
@@ -17,13 +17,13 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
-	[Desc("Allow deploying on specified charge to grant a condition for a specified duration.")]
-	public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo, IRulesetLoaded
+	[Desc("Grant a condition via player orders for a specified amount of time.")]
+	public class GrantChargedConditionOnToggleInfo : PausableConditionalTraitInfo, IRulesetLoaded
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs:21]: This uses "Toggle" but most of the fields below use "Activat(ed|ion)". Can we standardise on Activation if this must have a suffix? (I still don't think it needs one.)

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs (hunk 12 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs]: Charge-vs-not and Deploy-vs-activate are orthogonal features. The timed-condition-on-deploy feature should be either a feature of `GrantConditionOnDeploy` or a separate `GrantChargedConditionOnDeploy` trait.

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs (hunk 14 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs:21]: we do use the name toggle for toggle actions

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs (hunk 13 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs]: makes sense

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs (hunk 14 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs:21]: 🤷 
Screenshot 2024-12-21 at 15 54 05

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs:21]: ToggleState

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs [hunk 15]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
+		[Desc("The amount of charge that needs to be present to turn on the condition. " +
 			"If set to -1, threshold is set to full charge. " +
 			"If activated without full charge " + nameof(ConditionDuration) + " is percentally smaller.")]
 		public readonly int ChargeThreshhold = -1;
 
 		[Desc("How long (in ticks) should the condition stay active?")]
 		public readonly int ConditionDuration = 1;
 
-		[Desc("Can " + nameof(DeployedCondition) + " be canceled by followup deploy order?")]
+		[Desc("Can " + nameof(ActivatedCondition) + " be canceled turned off manually?")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs]:
```suggestion
		[Desc("Can " + nameof(ActivatedCondition) + " be turned off manually?")]
```

File: OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs [hunk 16]
```diff
@@ -35,38 +35,47 @@ public class GrantConditionOnDeployWithChargeInfo : PausableConditionalTraitInfo
...
 		public readonly bool CanCancelCondition = false;
 
+		[Desc("Should we interrupt the current activity")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs:49]: "... if not queued" is an important bit of information here.

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Conditions/GrantChargedConditionOnToggle.cs:49]: it makes it more confusing, in the same way my addition was

**Review by penev92 — APPROVED** [review]
The naming disagreements aside, thanks for the fix.

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/c835f93d6671935d8e112e09b16374a04c20a87a...efc1936f7186f09a500007b11a1db4df06bd1f68)

Commit on `prep-2411`:
https://github.com/OpenRA/OpenRA/commit/e0d893b7367f6929773b2ce54ab5fbed523e748c

**Comment by Mailaender** [comment]
The commit message has a typo.

**Comment by PunkPun** [comment]
It's a merged PR
